connection.go

```diff
@@ -93,3 +94,134 @@ func (e *EurekaConnection) UpdateApp(app *Application) {
 		}
 	}()
 }
+
+// AppUpdate is the outcome of an attempt to get a fresh snapshot of a Eureka
+// application's state, together with an error that may have occurred in that
+// attempt. If the Err field is nil, the App field will be non-nil.
+type AppUpdate struct {
+	App *Application
+	Err error
+}
+
+func sendAppUpdatesEvery(d time.Duration, produce func() AppUpdate, c chan<- AppUpdate, done <-chan struct{}) {
+	t := time.NewTicker(d)
+	defer t.Stop()
+	for {
+		select {
+		case <-done:
+			close(c)
+			return
+		case <-t.C:
+			// Drop attempted sends when the consumer hasn't received the last buffered update.
+			select {
+			case c <- produce():
+			default:
+			}
+		}
+	}
+}
+
+// ScheduleAppUpdates starts polling for updates to the Eureka application with
+// the given name, using the connection's configured polling interval as its
+// period. It sends the outcome of each update attempt to the returned channel,
+// and continues until the supplied done channel is either closed or has a value
+// available. Once done sending updates to the returned channel, it closes it.
+//
+// If prime is true, it sends at least one application update outcome to the
+// returned channel before returning.
+func (e *EurekaConnection) ScheduleAppUpdates(name string, prime bool, done <-chan struct{}) <-chan AppUpdate {
+	produce := func() AppUpdate {
+		app, err := e.GetApp(name)
+		return AppUpdate{app, err}
+	}
+	c := make(chan AppUpdate, 1)
+	if prime {
+		c <- produce()
+	}
+	go sendAppUpdatesEvery(time.Duration(e.PollInterval)*time.Second, produce, c, done)
+	return c
+}
+
+// An AppSource holds a periodically updated copy of a Eureka application.
+type AppSource struct {
+	m    *sync.RWMutex
+	app  *Application
+	done chan<- struct{}
+}
+
+// NewAppSource returns a new AppSource that offers a periodically updated copy
+// of the Eureka application with the given name, using the connection's
+// configured polling interval as its period.
+//
+// If prime is true, it waits for the first application update to complete
+// before returning, though it's possible that that first update attempt could
+// fail, so that a subsequent call to CopyLatestAppTo would return false.
+func (e *EurekaConnection) NewAppSource(name string, prime bool) *AppSource {
+	done := make(chan struct{})
+	updates := e.ScheduleAppUpdates(name, prime, done)
+	s := &AppSource{
+		done: done,
+	}
+	if prime {
+		if u := <-updates; u.Err != nil {
+			s.app = u.App
+		}
+	}
+	go func() {
+		for u := range updates {
+			if u.Err != nil {
```

Comment:
Between your two choices, I think I'd favor the first one. If the most recent update attempt wasn't successful it doesn't seem like we should try to hide that from the caller. 

As far as the self-preservation idea, I can definitely see it being useful but I'm a bit torn. It seems like something that might be use-case specific and should be implemented on the caller-side.